File: scripts/import-qualifications.js
const fs = require('fs');
const path = require('path');

const jsonData = JSON.parse(
  fs.readFileSync(
    path.join(__dirname, '../src/data/mancosa_qualifications.json'),
    'utf-8'
  )
);

const nqfLevelMap = {
  5: 'Higher Certificate',
  6: 'Advanced Certificate',
  7: 'Undergraduate Degree',
  8: 'Postgraduate Diploma',
  9: 'Masters Degree',
  10: 'Doctoral Degree'
};

const facultyMap = {
  'Business Administration': 'Business Management',
  'Public Administration': 'Public Management',
  'Commerce': 'Business and Commerce',
  'Education': 'Education',
  'Accounting': 'Accounting and Finance',
  'Marketing': 'Marketing and Communication',
  'Human Resource': 'Human Resources',
  'Information Technology': 'Information Technology',
  'Project Management': 'Project Management',
  'Supply Chain': 'Supply Chain Management'
};

function inferFaculty(title) {
  title = title.toLowerCase();
  if (title.includes('business administration')) return 'Business Management';
  if (title.includes('public administration')) return 'Public Management';
  if (title.includes('accounting')) return 'Accounting and Finance';
  if (title.includes('marketing')) return 'Marketing and Communication';
  if (title.includes('human resource')) return 'Human Resources';
  if (title.includes('information') || title.includes('technology')) return 'Information Technology';
  if (title.includes('project management')) return 'Project Management';
  if (title.includes('supply chain')) return 'Supply Chain Management';
  if (title.includes('financial')) return 'Accounting and Finance';
  if (title.includes('entrepreneurship')) return 'Business Management';
  if (title.includes('retail')) return 'Business Management';
  if (title.includes('tourism') || title.includes('hospitality')) return 'Tourism and Hospitality';
  if (title.includes('education')) return 'Education';
  if (title.includes('coding') || title.includes('artificial intelligence')) return 'Information Technology';
  if (title.includes('digital marketing')) return 'Marketing and Communication';
  return 'Business and Commerce';
}

function generateId(title) {
  const clean = title
    .replace(/Bachelor of Commerce/gi, 'BCOM')
    .replace(/Bachelor of Business Administration/gi, 'BBA')
    .replace(/Bachelor of Public Administration/gi, 'BPA')
    .replace(/Bachelor of Education/gi, 'BED')
    .replace(/Master of Business Administration/gi, 'MBA')
    .replace(/Master of Public Administration/gi, 'MPA')
    .replace(/Master of Commerce/gi, 'MCOM')
    .replace(/Doctor of Business Administration/gi, 'DBA')
    .replace(/Postgraduate Diploma/gi, 'PGDIP')
    .replace(/Advanced Diploma/gi, 'ADIP')
    .replace(/Advanced Certificate/gi, 'ACERT')
    .replace(/Higher Certificate/gi, 'HC')
    .replace(/Honours/gi, 'HONS')
    .replace(/\s+in\s+/gi, '_')
    .replace(/\s+/g, '_')
    .replace(/[^A-Z0-9_]/gi, '')
    .toUpperCase();

  return clean;
}

function parseDuration(durationStr) {
  if (!durationStr) return 1;
  const match = durationStr.match(/(\d+)/);
  return match ? parseInt(match[1]) : 1;
}

const qualifications = jsonData.map((q, index) => {
  const level = nqfLevelMap[q.NQF_Level] || 'Undergraduate Degree';
  const id = generateId(q.Title);
  const duration = parseDuration(q.Duration);
  const faculty = inferFaculty(q.Title);

  const careerPaths = q.CareerPaths || [];

  return {
    id,
    name: q.Title,
    level,
    nqf_level: q.NQF_Level,
    saqa_id: q.SAQA_ID || 'TBD',
    total_credits: q.Credits || 120,
    duration_years: duration,
    faculty,
    description: q.ShortDescription || q.About?.substring(0, 200) || '',
    coordinator: q.AcademicProgrammeManager || undefined,
    career_outcomes: careerPaths,
    programme_url: q.ProgrammeURL,
    saqa_url: q.SAQA_QualificationURL
  };
});

const tsContent = `/**
 * MANCOSA Qualification Catalogue
 * Comprehensive database of all MANCOSA qualifications and programmes
 * Auto-generated from mancosa_qualifications.json
 */

export interface Qualification {
  id: string;
  name: string;
  level: 'Higher Certificate' | 'Advanced Certificate' | 'Undergraduate Degree' | 'Advanced Diploma' | 'Postgraduate Diploma' | 'Honours Degree' | 'Masters Degree' | 'Doctoral Degree';
  nqf_level: 5 | 6 | 7 | 8 | 9 | 10;
  saqa_id: string;
  total_credits: number;
  duration_years: number;
  faculty: string;
  description: string;
  coordinator?: string;
  school?: string;
  career_outcomes?: string[];
  entry_requirements?: string;
  programme_url?: string;
  saqa_url?: string;
}

export const QUALIFICATIONS: Qualification[] = ${JSON.stringify(qualifications, null, 2)};

export function getQualificationById(id: string): Qualification | undefined {
  return QUALIFICATIONS.find(q => q.id === id);
}

export function getQualificationsByLevel(level: Qualification['level']): Qualification[] {
  return QUALIFICATIONS.filter(q => q.level === level);
}

export function getQualificationsByNQF(nqf_level: number): Qualification[] {
  return QUALIFICATIONS.filter(q => q.nqf_level === nqf_level);
}

export function getQualificationsByFaculty(faculty: string): Qualification[] {
  return QUALIFICATIONS.filter(q => q.faculty === faculty);
}

export function searchQualifications(query: string): Qualification[] {
  const lowerQuery = query.toLowerCase();
  return QUALIFICATIONS.filter(q =>
    q.name.toLowerCase().includes(lowerQuery) ||
    q.description.toLowerCase().includes(lowerQuery) ||
    q.faculty.toLowerCase().includes(lowerQuery)
  );
}

export function getAllFaculties(): string[] {
  return Array.from(new Set(QUALIFICATIONS.map(q => q.faculty))).sort();
}

export function getAllLevels(): Qualification['level'][] {
  return Array.from(new Set(QUALIFICATIONS.map(q => q.level)));
}
`;

fs.writeFileSync(
  path.join(__dirname, '../src/data/qualifications.ts'),
  tsContent,
  'utf-8'
);

console.log(`Successfully imported ${qualifications.length} qualifications`);
console.log('\nQualifications by level:');
const byLevel = qualifications.reduce((acc, q) => {
  acc[q.level] = (acc[q.level] || 0) + 1;
  return acc;
}, {});
console.log(byLevel);

console.log('\nQualifications by faculty:');
const byFaculty = qualifications.reduce((acc, q) => {
  acc[q.faculty] = (acc[q.faculty] || 0) + 1;
  return acc;
}, {});
console.log(byFaculty);

File: scripts/import-qualifications.py
#!/usr/bin/env python3
import json
import re
from pathlib import Path

# Read the JSON file
json_path = Path(__file__).parent.parent / 'src' / 'data' / 'mancosa_qualifications.json'
with open(json_path, 'r') as f:
    json_data = json.load(f)

nqf_level_map = {
    5: 'Higher Certificate',
    6: 'Advanced Certificate',
    7: 'Undergraduate Degree',
    8: 'Postgraduate Diploma',
    9: 'Masters Degree',
    10: 'Doctoral Degree'
}

def infer_faculty(title):
    title_lower = title.lower()
    if 'business administration' in title_lower:
        return 'Business Management'
    if 'public administration' in title_lower:
        return 'Public Management'
    if 'accounting' in title_lower:
        return 'Accounting and Finance'
    if 'marketing' in title_lower or 'digital marketing' in title_lower or 'social media' in title_lower or 'corporate communication' in title_lower:
        return 'Marketing and Communication'
    if 'human resource' in title_lower:
        return 'Human Resources'
    if 'information' in title_lower or 'technology' in title_lower or 'coding' in title_lower or 'artificial intelligence' in title_lower:
        return 'Information Technology'
    if 'project management' in title_lower:
        return 'Project Management'
    if 'supply chain' in title_lower or 'logistics' in title_lower:
        return 'Supply Chain Management'
    if 'financial' in title_lower or 'tax' in title_lower:
        return 'Accounting and Finance'
    if 'entrepreneurship' in title_lower or 'retail' in title_lower or 'family business' in title_lower:
        return 'Business Management'
    if 'tourism' in title_lower or 'hospitality' in title_lower or 'events' in title_lower:
        return 'Tourism and Hospitality'
    if 'education' in title_lower:
        return 'Education'
    if 'paralegal' in title_lower:
        return 'Legal Studies'
    return 'Business and Commerce'

def generate_id(title):
    replacements = [
        ('Bachelor of Commerce', 'BCOM'),
        ('Bachelor of Business Administration', 'BBA'),
        ('Bachelor of Public Administration', 'BPA'),
        ('Bachelor of Education', 'BED'),
        ('Master of Business Administration', 'MBA'),
        ('Master of Public Administration', 'MPA'),
        ('Master of Commerce', 'MCOM'),
        ('Doctor of Business Administration', 'DBA'),
        ('Postgraduate Diploma', 'PGDIP'),
        ('Advanced Diploma', 'ADIP'),
        ('Advanced Certificate', 'ACERT'),
        ('Higher Certificate', 'HC'),
        ('Honours', 'HONS'),
    ]

    clean = title
    for old, new in replacements:
        clean = re.sub(old, new, clean, flags=re.IGNORECASE)

    clean = re.sub(r'\s+in\s+', '_', clean, flags=re.IGNORECASE)
    clean = re.sub(r'\s+', '_', clean)
    clean = re.sub(r'[^A-Z0-9_]', '', clean, flags=re.IGNORECASE)

    return clean.upper()

def parse_duration(duration_str):
    if not duration_str:
        return 1
    match = re.search(r'(\d+)', duration_str)
    return int(match.group(1)) if match else 1

# Check if NQF Level 8 is Honours or Postgraduate Diploma
def determine_level_8_type(title):
    if 'honours' in title.lower():
        return 'Honours Degree'
    else:
        return 'Postgraduate Diploma'

qualifications = []
for q in json_data:
    nqf_level = q.get('NQF_Level')

    # Determine level
    if nqf_level == 8:
        level = determine_level_8_type(q['Title'])
    else:
        level = nqf_level_map.get(nqf_level, 'Undergraduate Degree')

    qual_id = generate_id(q['Title'])
    duration = parse_duration(q.get('Duration'))
    faculty = infer_faculty(q['Title'])

    career_paths = q.get('CareerPaths', [])

    description = q.get('ShortDescription', '')
    if not description and q.get('About'):
        description = q['About'][:200] + '...' if len(q['About']) > 200 else q['About']

    qualifications.append({
        'id': qual_id,
        'name': q['Title'],
        'level': level,
        'nqf_level': nqf_level,
        'saqa_id': q.get('SAQA_ID', 'TBD'),
        'total_credits': q.get('Credits', 120),
        'duration_years': duration,
        'faculty': faculty,
        'description': description,
        'coordinator': q.get('AcademicProgrammeManager') or None,
        'career_outcomes': career_paths,
        'programme_url': q.get('ProgrammeURL'),
        'saqa_url': q.get('SAQA_QualificationURL')
    })

# Generate TypeScript content
ts_content = '''/**
 * MANCOSA Qualification Catalogue
 * Comprehensive database of all MANCOSA qualifications and programmes
 * Auto-generated from mancosa_qualifications.json
 */

export interface Qualification {
  id: string;
  name: string;
  level: 'Higher Certificate' | 'Advanced Certificate' | 'Undergraduate Degree' | 'Advanced Diploma' | 'Postgraduate Diploma' | 'Honours Degree' | 'Masters Degree' | 'Doctoral Degree';
  nqf_level: 5 | 6 | 7 | 8 | 9 | 10;
  saqa_id: string;
  total_credits: number;
  duration_years: number;
  faculty: string;
  description: string;
  coordinator?: string;
  school?: string;
  career_outcomes?: string[];
  entry_requirements?: string;
  programme_url?: string;
  saqa_url?: string;
}

export const QUALIFICATIONS: Qualification[] = ''' + json.dumps(qualifications, indent=2) + ''';

export function getQualificationById(id: string): Qualification | undefined {
  return QUALIFICATIONS.find(q => q.id === id);
}

export function getQualificationsByLevel(level: Qualification['level']): Qualification[] {
  return QUALIFICATIONS.filter(q => q.level === level);
}

export function getQualificationsByNQF(nqf_level: number): Qualification[] {
  return QUALIFICATIONS.filter(q => q.nqf_level === nqf_level);
}

export function getQualificationsByFaculty(faculty: string): Qualification[] {
  return QUALIFICATIONS.filter(q => q.faculty === faculty);
}

export function searchQualifications(query: string): Qualification[] {
  const lowerQuery = query.toLowerCase();
  return QUALIFICATIONS.filter(q =>
    q.name.toLowerCase().includes(lowerQuery) ||
    q.description.toLowerCase().includes(lowerQuery) ||
    q.faculty.toLowerCase().includes(lowerQuery)
  );
}

export function getAllFaculties(): string[] {
  return Array.from(new Set(QUALIFICATIONS.map(q => q.faculty))).sort();
}

export function getAllLevels(): Qualification['level'][] {
  return Array.from(new Set(QUALIFICATIONS.map(q => q.level)));
}
'''

# Write the TypeScript file
output_path = Path(__file__).parent.parent / 'src' / 'data' / 'qualifications.ts'
with open(output_path, 'w') as f:
    f.write(ts_content)

print(f"Successfully imported {len(qualifications)} qualifications")
print("\nQualifications by level:")
by_level = {}
for q in qualifications:
    level = q['level']
    by_level[level] = by_level.get(level, 0) + 1
for level, count in sorted(by_level.items()):
    print(f"  {level}: {count}")

print("\nQualifications by faculty:")
by_faculty = {}
for q in qualifications:
    faculty = q['faculty']
    by_faculty[faculty] = by_faculty.get(faculty, 0) + 1
for faculty, count in sorted(by_faculty.items()):
    print(f"  {faculty}: {count}")
